ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/ClientTest.java

```diff
@@ -0,0 +1,308 @@
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.client;
+
+import static io.confluent.ksql.test.util.AssertEventually.assertThatEventually;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+import io.confluent.ksql.api.BaseApiTest;
+import io.confluent.ksql.api.client.impl.ClientImpl;
+import io.confluent.ksql.api.client.impl.ClientOptionsImpl;
+import io.confluent.ksql.api.server.PushQueryId;
+import io.confluent.ksql.parser.exception.ParseFailedException;
+import io.vertx.core.Context;
+import io.vertx.ext.web.client.WebClient;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.junit.Test;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public class ClientTest extends BaseApiTest {
+
+  @SuppressWarnings("unchecked")
+  protected static final List<String> DEFAULT_COLUMN_NAMES = BaseApiTest.DEFAULT_COLUMN_NAMES.getList();
+  @SuppressWarnings("unchecked")
+  protected static final List<String> DEFAULT_COLUMN_TYPES = BaseApiTest.DEFAULT_COLUMN_TYPES.getList();
+  protected static final Map<String, Object> DEFAULT_PUSH_QUERY_REQUEST_PROPERTIES =
+      BaseApiTest.DEFAULT_PUSH_QUERY_REQUEST_PROPERTIES.getMap();
+
+  protected Context context;
+  protected Client client;
+
+  @Override
+  public void setUp() {
+    super.setUp();
+
+    context = vertx.getOrCreateContext();
+  }
+
+  @Override
+  protected WebClient createClient() {
+    // Use Java client for these tests, rather than a vanilla WebClient
+    this.client = createJavaClient();
+
+    return null;
+  }
+
+  @Override
+  protected void stopClient() {
+    if (client != null) {
+      try {
+        client.close();
+      } catch (Exception e) {
+        log.error("Failed to close client", e);
+      }
+    }
+  }
+
+  @Test
+  public void shouldStreamQueryAsync() throws Exception {
+    // When
+    final CompletableFuture<QueryResult> cf = client.streamQuery(DEFAULT_PUSH_QUERY, DEFAULT_PUSH_QUERY_REQUEST_PROPERTIES);
+    assertThatEventually(cf::isDone, equalTo(true));
+    assertThat(cf.isCompletedExceptionally(), equalTo(false));
```

Comment:
This line and then one above could be replace with a cf.get() to return the result of the cf synchronously. This makes tests like this a bit easier to follow imho. cf.get() will cause exception to be thrown if it completed exceptionally.

Reply:
Also cf.get() will block the calling thread until the cf is complete, which is more efficient than polling with assertThatEventually.

Reply:
Much cleaner, thanks for the suggestion! Made the changes.